### DISCUSSION TITLE
[Feature] skywalking as a platform and multi tenant

### DISCUSSION THREAD

**Discussion by chenhaipeng** [apache/skywalking#10329] [Ideas]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

I want to use skywalking as a platform and use tenant id to isolate the data of each business. My solution is to use receiver-sharing-server.authentication to identify the data of different businesses and increase the storage ID. I don’t know how other people deal with it . Does the community have a recommended solution?

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
Hi, multiple tenants concept is a project level thing. It should be very serious.
What you proposed is too simple to uncover a lot of things. We need a complete proposal if you are going to work on this.
1. What is the definition of tenant.
2. How and where define a tenant? Such as is there management feature for it?
3. If the data is out of SkyWalking native agent, how to define it.
4. How tenant affects the storage data and formats.
5. Which isolation level among tenants. This is very important, we may face a lot of potential CVE reports due to tenant boundary breaking is a serious issue.
6. How UI works with this new concept? Do you or anyone one on your team promise to implement quickly, nd cover maintainance in the future?
7. How our export, alerting, MAL, OAL, LAL work with tenant concept?

This is a simple version of FAQs if you want to deeply dive into this topic.

**Comment by wu-sheng** [comment]
If you ask how SkyWalking did, it is no such concept.

**Reply by chenhaipeng** [reply]
1. A tenant is just an identity, and different businesses correspond to different identities. Currently, Skywalking does not support different business data isolation
2. My initial idea is to use a table to maintain access information, access A, access B, etc.
3. The access must carry the configured access token, which is the service report
4. storage has one more field, tenantId distinguishes different access

**Reply by wu-sheng** [reply]
That is not just a new field. Have you considered when two tenants have services in the same name?

**Reply by chenhaipeng** [reply]
6. According to the data requested by the tenantId, the UI can pull the access permission data, and each request carries the tenantId to distinguish the corresponding stored tenantId field

**Reply by chenhaipeng** [reply]
7.put tenantId Integrate into the strategy of generating serviceId, if MAL, OAL, LAL are associated with serviceId, then there is no need to modify

**Reply by chenhaipeng** [reply]
> That is not just a new field. Have you considered when two tenants have services in the same name?

see number7.  put the tenantId to generate serviceId,  mean:seviceName + tenantId

**Reply by wu-sheng** [reply]
What stategy? You can't break the whole ecosystem add a required field.

**Reply by wu-sheng** [reply]
We need a practical solution and design doc. This seems to be the biggest change since v6.

**Reply by chenhaipeng** [reply]
> > That is not just a new field. Have you considered when two tenants have services in the same name?
> 
> see number7. put the tenantId to generate serviceId, mean:seviceName + tenantId

<img width="1122" alt="image" src="https://user-images.githubusercontent.com/6678954/216908951-42d1bad4-bc6c-4b68-9c05-b67a3033a703.png">

**Reply by chenhaipeng** [reply]
> We need a practical solution and design doc. This seems to be the biggest change since v6.

Yes, this is indeed a very big change. If Skywalking is used as a platform, this function is necessary

**Reply by wu-sheng** [reply]
> > > That is not just a new field. Have you considered when two tenants have services in the same name?
> 
> > 
> 
> > see number7. put the tenantId to generate serviceId, mean:seviceName + tenantId
> 
> 
> 
> <img width="1122" alt="image" src="https://user-images.githubusercontent.com/6678954/216908951-42d1bad4-bc6c-4b68-9c05-b67a3033a703.png">
> 
> 

You could do this as a private implementation. But in reality, all integrations(agents, querys) and all codes(relying on ID) are down.

**Reply by wu-sheng** [reply]
> > We need a practical solution and design doc. This seems to be the biggest change since v6.
> 
> 
> 
> Yes, this is indeed a very big change. If Skywalking is used as a platform, this function is necessary

SkyWalking is an open source software, we are not a platform. So, your statement is not suitable, instead you build a platform using SkyWalking, that is it.

**Reply by chenhaipeng** [reply]
en, This is a destructive function，I modified AuthenticationInterceptor.java to inject tenantId,
I modified AuthenticationInterceptor.java to inject tenantId. In fact, this class does not do this.

**Reply by wu-sheng** [reply]
About granting access, you either pick RABC or NGAC models to manage, rather than randomly picking up fields/tables.

**Reply by wu-sheng** [reply]
> en, This is a destructive function，I modified AuthenticationInterceptor.java to inject tenantId,
> 
> I modified AuthenticationInterceptor.java to inject tenantId. In fact, this class does not do this.

That is a naive change, only surviving in your private use case.
SkyWalking has dozens of data sources, what you proposed at agent side is actually not working from my perspective. 

And trust a client ID without decoding is a common security issue.

**Reply by chenhaipeng** [reply]
UI -> UI-proxy ->RABC -> GraphQL

**Reply by chenhaipeng** [reply]
> > en, This is a destructive function，I modified AuthenticationInterceptor.java to inject tenantId,
> > I modified AuthenticationInterceptor.java to inject tenantId. In fact, this class does not do this.
> 
> That is a naive change, only surviving in your private use case. SkyWalking has dozens of data sources, what you proposed at agent side is actually not working from my perspective.
> 
> And trust a client ID without decoding is a common security issue.

The client ID is just a token, you can sign the clientId, Skywalking OAP receives the clientId, whether to trust or decode it is another question

**Reply by wu-sheng** [reply]
> UI -> UI-proxy ->RABC -> GraphQL

If RABC is not in OAP directly accessing, this could be another CVE

This is what I mean it is a big change.

If you want to do, let's do based on an official doc(google doc) with all details. Why, how to change, how to verify, security check list, 3rd party ecosystem status(breaking?)

You need to be very detail very careful. And every breaking decision you made may bring you -1 for voting.

**Reply by chenhaipeng** [reply]
> > UI -> UI-proxy ->RABC -> GraphQL
> 
> If RABC is not in OAP directly accessing, this could be another CVE
> 
> This is what I mean it is a big change.
> 
> If you want to do, let's do based on an official doc(google doc) with all details. Why, how to change, how to verify, security check list, 3rd party ecosystem status(breaking?)
> 
> You need to be very detail very careful. And every breaking decision you made may bring you -1 for voting.

Yes, at present, only a small number of appeals require this feature to be used as a platform. Let’s discuss the necessity first.

**Reply by wu-sheng** [reply]
For me, I don't know either. Tenancy is always a hot topic.

**Comment by chenhaipeng** [comment]
I have simply implemented the function of distinguishing tenants. If the community needs this feature, I am happy to help submit it

**Reply by Superskyyy** [reply]
AFAIK, there have been many users asking for this feature, but still gotta be very careful as this generally is a easy place to get dozens of CVEs.
